state-commitments/optimized/history_commitment.go

```diff
@@ -0,0 +1,489 @@
+package optimized
+
+import (
+	"errors"
+	"fmt"
+	"math"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+var emptyHash = common.Hash{}
+
+type Commitment struct {
+	Height         uint64
+	Merkle         common.Hash
+	FirstLeaf      common.Hash
+	LastLeafProof  []common.Hash
+	FirstLeafProof []common.Hash
+	LastLeaf       common.Hash
+}
+
+func NewCommitment(leaves []common.Hash, virtual uint64) (*Commitment, error) {
+	if len(leaves) == 0 {
+		return nil, errors.New("must commit to at least one leaf")
+	}
+	comm := NewCommitter()
+	root, err := comm.ComputeRoot(leaves, virtual)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: Generate the inclusion proofs necessary.
+	return &Commitment{
+		Height:    virtual,
+		Merkle:    root,
+		FirstLeaf: leaves[0],
+		LastLeaf:  leaves[len(leaves)-1],
+	}, nil
+}
+
+type HistoryCommitter struct {
+	lastLeafFillers []common.Hash
+	keccak          crypto.KeccakState
+}
+
+func NewCommitter() *HistoryCommitter {
+	return &HistoryCommitter{
+		lastLeafFillers: make([]common.Hash, 0),
+		keccak:          crypto.NewKeccakState(),
+	}
+}
+
+func (h *HistoryCommitter) ComputeRoot(leaves []common.Hash, virtual uint64) (common.Hash, error) {
+	if len(leaves) == 0 {
+		return common.Hash{}, nil
+	}
+	// Called with 0 limit first to compute the last leaf fillers for the commitment.
+	copiedLeaves := make([]common.Hash, len(leaves))
+	for i, leaf := range leaves {
+		var copied common.Hash
+		copy(copied[:], leaf[:])
+		copiedLeaves[i] = copied
+	}
+	limit := nextPowerOf2(virtual)
+	_, err := h.computeVirtualSparseTree(copiedLeaves, virtual, 0)
+	if err != nil {
+		return common.Hash{}, err
+	}
+	return h.computeVirtualSparseTree(leaves, virtual, limit)
+}
+
+func (h *HistoryCommitter) GeneratePrefixProof(prefixIndex uint64, leaves []common.Hash, virtual uint64) ([]common.Hash, []common.Hash, error) {
+	prefixExpansion, proof, err := h.prefixAndProof(prefixIndex, leaves, virtual)
+	if err != nil {
+		return nil, nil, err
+	}
+	prefixExpansion = trimTrailingZeroHashes(prefixExpansion)
+	proof = trimZeroes(proof)
+	return prefixExpansion, proof, nil
+}
+
+// computeSparseTree returns the htr of a hashtree with the given leaves and
+// limit. Any non-allocated leaf is filled with the passed zeroHash of depth 0.
+// Recursively, any non allocated intermediate layer at depth i is filled with
+// the passed zeroHash.
+// limit is assumed to be a power of two which is higher or equal than the
+// length of the leaves.
+// fillers is assumed to be precomputed to the necessary limit.
+//
+// Zero allocations
+// Computes O(len(leaves)) hashes.
+func (h *HistoryCommitter) computeSparseTree(leaves []common.Hash, limit uint64, fillers []common.Hash) (common.Hash, error) {
+	if len(leaves) == 0 {
+		return common.Hash{}, nil
+	}
+	if limit < 2 {
```

Comment:
This would be wrong in case `limit == 0`, this doesn't happen as this function is not exported and it's only called with a power of two, however at least a comment would be good here.